OP-Supervisor: Expand EntryType to include Replacement Type ; Use Replacement Type to trigger Replacements for Syncing Nodes

axelKingsley opened the pull request:
- Expands the `invalidated` bool to use the full `EntryType` byte
- Adds `ReplacementV0` as a new Entry Type
- Adds unit testing for all Entry Type parsing, including encode/decode and invalid Types
- Marks the EntryType as a `ReplacementV0` when the `invalidated` block hash is provided to `addLink`
- Wires Replacement checks through the backend
- Adds a new `ReplaceIfNeeded` check for the Managed Node
  - If there is ever an `ErrConflict` when the node reports its new derived block, EITHER:
    - The Node is genuinely out of sync and needs to initate a reset, OR
    - The Node is encountering a natural requirement to invalidate+replace the block they just derived
  - So, we check if a replacement is needed first *before* starting down the reset route
- Updates the `LastSeenSafe` and `LastSeenUnsafe` when a replacement comes in from the Managed Node, and triggers a `ResetIfInconsistent` check. This way if the node is corrupted *after* a replacement, we will still notice. This also makes it so nodes which just did a replacement won't get incorrectly flagged as inconsistent
- Fixes a channel closure bug when Managed Nodes close

axelKingsley committed:
Expand EntryType system ; Include Replacement Type in AddLink

axelKingsley committed:
Wire Replacement Check to Managed Node and Handle Sync Replacements

protolambda commented on op-supervisor/supervisor/backend/db/fromda/db.go:
```diff
@@ -137,6 +137,19 @@ func (db *DB) Invalidated() (pair types.DerivedBlockSealPair, err error) {
 	}, nil
 }
 
+func (db *DB) Replacement(derived eth.BlockID) error {
+	db.rwLock.RLock()
+	defer db.rwLock.RUnlock()
+	_, link, err := db.derivedNumToFirstSource(derived.Number)
```
We shouldn't add APIs that take a `eth.BlockID` argument and only use the block-number, since that falsely implies that the block-hash is being checked, which it is not here. Let's check that the link has a block-hash that matches the derived block hash.

axelKingsley replied:
I think this is a misleading parameter name that I should change -- If a caller is asking if a block is a replacement block, they may not yet have the correct derived, and so we shouldn't expect it to match. The caller has likely derived an invalid block and is now attempting to find out if the item which *is* in the database (which doesn't match) is a replacement.

Since we can't actually do a block check here, I'll reframe the API to `blockNum uint64`

protolambda replied:
We can do a block check here, the L2 blockhash is part of the derived side of the link. Doing the blockhash check can remove surprises from the caller, by making the check not report an invalid thing when there's some reorg.

protolambda commented on op-supervisor/supervisor/backend/db/fromda/db_test.go:
```diff
@@ -1140,9 +1140,9 @@ func TestLookupDetectIfCorruptDB(t *testing.T) {
-				invalidated: false,
+				source:    l1Block2,
+				derived:   l2Block2,
+				entryType: InvalidatedFromV0,
```
Why change this to invalidated? Shouldn't it be a `SourceV0`?

axelKingsley replied:
A miss during refactor, indeed an error. Will fix, thanks

protolambda commented on op-supervisor/supervisor/backend/db/fromda/update.go:
```diff
@@ -226,15 +230,17 @@ func (db *DB) addLink(source eth.BlockRef, derived eth.BlockRef, invalidated com
 				return fmt.Errorf("derived block %s conflicts with known derived block %s at same height: %w",
 					derived, lastDerived, types.ErrConflict)
 			}
+			// invalidRef is used and references pass, so this is actually a replacement block
+			link.entryType = ReplacementV0
```
This comment makes no sense to me, can you explain it?
This happens in the `else` of the `invalidRef != (common.Hash{})`, so `invalidRef` is *not* being used here.

And we just checked that we're deriving the same L2 block-hash as last known, so the L2 block is not being replaced, but the L1 block source is changing.

Is this `link.entryType = ReplacementV0` meant to go in a different place?

axelKingsley replied:
Ah, my bad, this was meant to be located ~5 lines higher, when the ref is confirmed to match the `lastDerived`. Not sure why I ended up putting it lower.

In general I've had some issues parsing the full logic of `addLink` since it's quite long and has a lot of branches. Anyway, my mistake, will fix.

axelKingsley replied:
I am going over the function more and I'm thinking I'll move this check to the top of the function, next to the other type settings. I figure we know it's a replacement if the `invalid` is set but doesn't match the `derived` provided.

I think there's a function comment somewhere that was misleading, about how `invalid` must always match derived. If I see that again I'll try to clear it up.

axelKingsley replied:
Ya, I made this a much nice type switch at the top of the function, and got out of the way of the consistency checks. Much cleaner IMO. And made the comment a little more clear about the different modalities with `invalidRef`

protolambda commented on op-supervisor/supervisor/backend/syncnode/node.go:
```diff
@@ -420,6 +434,34 @@ func (m *ManagedNode) Close() error {
+func (m *ManagedNode) replaceIfNeeded() bool {
+	ctx, cancel := context.WithTimeout(m.ctx, internalTimeout)
+	defer cancel()
+	err := m.backend.IsReplacement(ctx, m.chainID, m.lastNodeLocalSafe)
+	if err != nil {
+		m.log.Info("Replacement block detected, initiating replacement",
+			"lastNodeLocalSafe", m.lastNodeLocalSafe,
+			"err", err)
```
Why `err != nil`? Wouldn't it also return an error if the block in the DB is an invalidated block?
I think it returns an error only if the block is *not* a replacement, the opposite of what we're trying to do here.
Can you add an action test to show that a new node can sync, after a prior replacement?

axelKingsley replied:
Ah yeah you're right, I'm returning `nil` if it *is* a replacement, and then I flipped my logic here.

protolambda commented:
The inner DB change to add the link-entry type looks good, but the outer logic that works with the replacement does not seem right

axelKingsley committed:
Address comments ; break unit test

axelKingsley commented:
I've addressed comments and added to some existing test, as I realized there are opportunities already to check if a block is replaced, which revealed there may be a pre-existing issue with invalid blocks remaining in the database.

Will need to look into that.

EDIT: this is an edge case with how invalid blocks are recorded when the L1 that replaces it is later. It means the current `Replaces` check is not sufficient for this edge case.

EDIT2: resolved

axelKingsley committed:
Fix Replacement ; Remove Syncnode use of Replacement

axelKingsley commented:
I've removed the code in `syncnode` which uses the Replacment markers, as that part needs more cook time. This PR now features the code to write replacement-type blocks, and also includes a `Replacement` check function which is currently used in `fromda` tests only.

axelKingsley commented:
I don't think we actually need to track this -- discrepancy between the Local and Cross Safe databases will reveal any gaps which indicate a replacement.
